### PR TITLE
[api] Add methods to find min/maximum value in QgsRasterBlock

### PR DESCRIPTION
**Pull request by nyalldawson**

### CODE REVIEW

**Comment by JanCaha** [comment]
Nice, this will be really useful :+1:

**Comment by nyalldawson** [comment]
@rouault how's this look to you? Any ideas for faster/better approaches?

File: src/core/raster/qgsrasterblock.cpp [hunk 1]
```diff
@@ -872,3 +872,293 @@ QRect QgsRasterBlock::subRect( const QgsRectangle &extent, int width, int height
...
   return subRect;
 }
+
+template<class T> bool findMaximum( const void *data, bool hasNoData, double noDataValue, int width, int height, double &maximum, std::size_t &offset )
+{
+  const T *castData = static_cast< const T * >( data );
+  auto end = castData + static_cast< std::size_t >( width ) * static_cast< std::size_t >( height );
+  if ( !hasNoData )
+  {
+    const auto maxElem = std::max_element( castData, end );
```
Comment (rouault) [inline src/core/raster/qgsrasterblock.cpp:882]: you might run into trouble with float and double when they are NaN: https://stackoverflow.com/a/35890119/23076275 . Same for std::min_element and presumably minmax_element

**Review by rouault — COMMENTED** [review]
LGTM generally, but the comment.
Ideally QGIS wouldn't have to do that itself and this would sit as a GDAL utility function, since that might be useful here...

**Comment by rouault** [comment]
Regarding my GDAL mention, we do have optimized min/max computation in GDALComputeRasterMinMax(), at the band level, for common data types such as Byte and UInt16. If you were not interested in the place where the min/max is met, wrapping a raw buffer as a GDALRasterBand could be a way of leveraging that. But, for just a block, I'm not sure the setup cost of creating a MEM rasterband wouldn't hurt performance. We could potentially move that code to a utility function, that would just take the start pointer, data type, number of elements and nodata value.

File: src/core/raster/qgsrasterblock.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/raster/qgsrasterblock.cpp:882]: Eww, that's painful! I can't think of any alternative but to fallback to the non-std branch for double/float types then...

**Comment by nyalldawson** [comment]
@rouault 

> Regarding my GDAL mention, we do have optimized min/max computation in GDALComputeRasterMinMax(), at the band level, for common data types such as Byte and UInt16. If you were not interested in the place where the min/max is met

Finding the row/col is actually the main use case here (it was inspired by that recent gdal-dev thread, and I plan on exposing as a processing tool after the API is finalised).

> We could potentially move that code to a utility function, that would just take the start pointer, data type, number of elements and nodata value.

Do you mean a utility function in gdal?

File: src/core/raster/qgsrasterblock.cpp (hunk 1 above)
Comment (rouault) [reply src/core/raster/qgsrasterblock.cpp:882]: Create a specific branch when ``if constexpr (std::is_same<T, float>::value || std::is_same<T, double>::value)`` that uses a custom NaN friendly comparison function

**Comment by rouault** [comment]
> Finding the row/col is actually the main use case here (it was inspired by that recent gdal-dev thread, and I plan on exposing as a processing tool after the API is finalised).

yes, got it.

>  Do you mean a utility function in gdal?

yes. Basically my message: this would be wonderful to move all of this in GDAL. But I understand this would imply a turn around of 6 months +.

**Comment by rouault** [comment]
@nyalldawson https://github.com/OSGeo/gdal/pull/11199 provides a gdal_minmax_element.hpp public header, that can also be vendored in QGIS, to find the min/max elements in a buffer. It properly takes into account NaN for float/double. And there's an optimized SSE2 version for float/double in the no nodata case.

**Comment by nyalldawson** [comment]
@rouault is [2d1f5fd](https://github.com/qgis/QGIS/pull/59285/commits/2d1f5fdecbc42fd8473d3daedf8976a7303ec8e2) the correct approach to use the gdal method?

**Comment by rouault** [comment]
> is [2d1f5fd](https://github.com/qgis/QGIS/pull/59285/commits/2d1f5fdecbc42fd8473d3daedf8976a7303ec8e2) the correct approach to use the gdal method?

yes that looks good. You'll also need https://github.com/rouault/gdal/commit/4e4ff4105000ca4926fc164b09738c8f4a635b73 to fix the build for GDAL < 3.7

**Comment by rouault** [comment]
> You'll also need [rouault/gdal@4e4ff41](https://github.com/rouault/gdal/commit/4e4ff4105000ca4926fc164b09738c8f4a635b73)

edit: that should be [rouault/gdal@9164d9d55d2d81c4699ce3ee2fb3feaec998b86b](https://github.com/rouault/gdal/commit/9164d9d55d2d81c4699ce3ee2fb3feaec998b86b)

File: src/core/raster/qgsrasterblock.cpp [hunk 2]
```diff
@@ -872,3 +876,63 @@ QRect QgsRasterBlock::subRect( const QgsRectangle &extent, int width, int height
...
+  {
+    minimum = std::numeric_limits<double>::quiet_NaN();
+    maximum = std::numeric_limits<double>::quiet_NaN();
+    return false;
+  }
+
+  const auto offset = qgis_gdal::minmax_element( mData, static_cast<std::size_t>( mWidth ) * static_cast< std::size_t>( mHeight ),
+                      QgsGdalUtils::gdalDataTypeFromQgisDataType( mDataType ), mHasNoDataValue, mNoDataValue );
+  std::size_t minOffset = offset.first;
+  std::size_t maxOffset = offset.second;
```
Comment (rouault) [inline src/core/raster/qgsrasterblock.cpp:927]: could be simplified as:
```suggestion
  const auto [minOffset, maxOffset] = qgis_gdal::minmax_element( mData, static_cast<std::size_t>( mWidth ) * static_cast< std::size_t>( mHeight ),
                      QgsGdalUtils::gdalDataTypeFromQgisDataType( mDataType ), mHasNoDataValue, mNoDataValue );
```

**Comment by nyalldawson** [comment]
Thanks @rouault , I've updated accordingly

**Review by rouault — APPROVED** [review]

**Pull request merged**